### PR TITLE
Make time errors more explicit when setting the ":next" Testday

### PR DESCRIPTION
**Pull request by xabolcs**
Fix for #74.

### CODE REVIEW

**Comment by xabolcs** [comment]
Example talk with my testdaybot named xabolcs_tdb:

```
[23:46:38]	xabolcs	:next
[23:46:39]	xabolcs_tdb	Need some help? :next <start as YYYY-MM-DDThh:mmZ> <end as YYYY-MM-DDThh:mmZ> <etherpad> <topic>: set next Test Day
[23:49:41]	xabolcs	:next 2014-12-13T23:30Z 2013-12-13T23:45Z etherpad topic
[23:49:41]	xabolcs_tdb	Please use valid dates.
[23:49:41]	xabolcs_tdb	Start time was set after End time.
[23:50:09]	xabolcs	:next 2014-12-13T23:30Z 2014-12-13T23:45Z etherpad topic
[23:50:10]	xabolcs_tdb	Please use valid dates.
[23:50:10]	xabolcs_tdb	Start time is set in the past.
[23:50:29]	xabolcs	:next 2015-12-13T23:30Z 2014-12-13T23:45Z etherpad topic
[23:50:30]	xabolcs_tdb	Please use valid dates.
[23:50:30]	xabolcs_tdb	Start time was set after End time.
[23:50:35]	xabolcs	:next 2015-12-13T23:30Z 2015-12-13T23:45Z etherpad topic
[23:50:36]	xabolcs_tdb	Next Test Day's start is Sun, 13 Dec 2015 23:30:00 GMT
[23:50:36]	xabolcs_tdb	Next Test Day's end is Sun, 13 Dec 2015 23:45:00 GMT
[23:50:36]	xabolcs_tdb	Next Test Day's etherpad is etherpad
[23:50:36]	xabolcs_tdb	Next Test Day's topic is topic
```

**Comment by xabolcs** [comment]
Interestingly there was a message in the channel after setting up a test day. Was it expected?

```
[23:50:36]	xabolcs_tdb	The topic testday starts now. For details, see etherpad
```

File: bot.js [hunk 1]
```diff
@@ -418,6 +418,11 @@ client.addListener('pm', function(from, message) { // private messages to bot
               client.say(from, "Next Test Day's topic is " + testDay.topic);
             } else {
               client.say(from, "Please use valid dates.");
+              if (testDay.end < testDay.start) {
+                client.say(from, "Start time was set after End time.");
+              } else {
```
Comment (xabolcs) [inline bot.js]: Other option here is using `if` to print all errors at once.

@whimboo, should I go that way? Currently only one error is printed.

An advanced solution would be to count all the errors and print a it like "Your specified a time interval which has the following 2 errors: the start time is after end time, the start time is in the past."

Thoughts?

**Comment by xabolcs** [comment]
This is ready to review!

File: bot.js (hunk 1 above)
Comment (whimboo) [inline bot.js]: You could use a string variable for the `client.say` call in line 420, and then just combine the text for the individual error message like "Please use valid dates: %error%". That way we also only send a single message to the user.

I don't think the extended logic is that more helpful right now. Personally I would move those sanity checks for times up before line 410, so we do not have to check for the dates twice but return early.

Comment (xabolcs) [inline bot.js]: > Personally I would move those sanity checks for times up before line 410, so we do not have to check for the dates twice but return early.

I happily do this move in this PR, should I?

Comment (whimboo) [inline bot.js]: That would be perfect.

**Comment by xabolcs** [comment]
Pushed commits to PR's branch including 45f4fc910c69f42ac23a289c0a6a587e7fe061dc.
Ready to review.

Please see the following conversation!

```

URL 	ircs://moznet/xabolcs_tdb,isnick 	Connected via 	fripp.mozilla.org
Conversation with xabolcs_tdb <nodebot@moz-fci6nk.catv.broadband.hu>
nodeJS IRC client
xabolcs_tdb
[13:54:06]	[INFO]	Messages Cleared.
[13:54:11]	xabolcs	:next
[13:54:12]	xabolcs_tdb	Need some help? :next <start as YYYY-MM-DDThh:mmZ> <end as YYYY-MM-DDThh:mmZ> <etherpad> <topic>: set next Test Day
[13:55:46]	xabolcs	:next 2014-03-04T11:00Z 2014-03-04T11:30Z https://etherpad.location.url TestDayBot Test Day
[13:55:46]	xabolcs_tdb	Please use valid dates: start time is set in the past.
[13:56:01]	xabolcs	:next 2014-03-04T11:00Z 2014-03-04T10:30Z https://etherpad.location.url TestDayBot Test Day
[13:56:02]	xabolcs_tdb	Please use valid dates: start time is set after End time, start time is set in the past.
[13:56:24]	xabolcs	:next 2015-03-04T11:00Z 2014-03-04T10:30Z https://etherpad.location.url TestDayBot Test Day
[13:56:24]	xabolcs_tdb	Please use valid dates: start time is set after End time.
[13:56:43]	xabolcs	:next 2015-03-04T11:00Z 2015-03-04T11:30Z https://etherpad.location.url TestDayBot Test Day
[13:56:43]	xabolcs_tdb	Next Test Day's start is Wed, 04 Mar 2015 11:00:00 GMT
[13:56:43]	xabolcs_tdb	Next Test Day's end is Wed, 04 Mar 2015 11:30:00 GMT
[13:56:43]	xabolcs_tdb	Next Test Day's etherpad is https://etherpad.location.url
[13:56:43]	xabolcs_tdb	Next Test Day's topic is TestDayBot Test Day
```

File: bot.js [hunk 2]
```diff
@@ -402,22 +402,32 @@ client.addListener('pm', function(from, message) { // private messages to bot
...
+            if (startTime < Date.now()) {
+              dateErrors.push("start time is set in the past");
+            }
+            // if the start and end dates appear valid, set the test day date
+            if (dateErrors.length == 0) {
               if (timerID !== 0) {
                 clearTimeout(timerID);
               }
+              testDay.start = new Date(command[1]);
+              testDay.end = new Date(command[2]);
```
Comment (xabolcs) [inline bot.js]: Ahh, these should be `startTime` and `endTime`.

**Comment by xabolcs** [comment]
This is ready to review!

File: bot.js [hunk 3]
```diff
@@ -402,22 +402,32 @@ client.addListener('pm', function(from, message) { // private messages to bot
           client.say(from, "Test Day in progress and scheduled to end " + testDay.end);
         } else {
           if (cmdLen >= 5) {
-            testDay.start = new Date(command[1]);
-            testDay.end = new Date(command[2]);
-            testDay.etherpad = command[3];
-            testDay.topic = command.slice(4, cmdLen).join(" ");
-            // if the start and end dates appear valid, set the test date
-            if ((testDay.end > testDay.start) && (testDay.start > Date.now())) {
+            var startTime, endTime, dateErrors = [];
```
Comment (whimboo) [inline bot.js]: Do not declare multiple variables in a single line. So move the declaration for startTime and endTime to the lines 406 and 407.

File: bot.js [hunk 4]
```diff
@@ -402,22 +402,32 @@ client.addListener('pm', function(from, message) { // private messages to bot
...
+            startTime = new Date(command[1]);
+            endTime = new Date(command[2]);
+            if (endTime < startTime) {
+              dateErrors.push("start time is set after End time");
```
Comment (whimboo) [inline bot.js]: nit: Please start the message with a capital letter and lowercase `end`.

File: bot.js [hunk 5]
```diff
@@ -402,22 +402,32 @@ client.addListener('pm', function(from, message) { // private messages to bot
...
               timerID = setTimeout(updateTestDayData, testDay.start - Date.now());
               client.say(from, "Next Test Day's start is " + testDay.start.toUTCString());
               client.say(from, "Next Test Day's end is " + testDay.end.toUTCString());
               client.say(from, "Next Test Day's etherpad is " + testDay.etherpad);
               client.say(from, "Next Test Day's topic is " + testDay.topic);
-            } else {
-              client.say(from, "Please use valid dates.");
+            }
+            else {
+              client.say(from, "Please use valid dates: " + dateErrors.join(", ") + ".");
```
Comment (whimboo) [inline bot.js]: I would suggest to print each failure in its own line, which makes it easier to read.

Comment (xabolcs) [inline bot.js]: Sorry, I don't understand.
How does this correlate to your [previous][1] comment?
> You could use a string variable for the client.say call in line 420, and then just combine the text for the individual error message like "Please use valid dates: %error%". That way we also only send a single message to the user.


[1]: #discussion_r25408527

Comment (whimboo) [inline bot.js]: Simply join the array entries with `\n` instead.

Comment (xabolcs) [inline bot.js]: Thanks!

**Comment by xabolcs** [comment]
This is ready to review up to commit b14a8a01ac9f1ec0e277fd3f35728c542fa906f6.

**Comment by whimboo** [comment]
This looks fine! I got it merged as https://github.com/mozilla/testdaybot/commit/5a749f976b3a0b79e920d0bc5bdde439f4b37f0a. Thanks for your contribution.

**Pull request closed without merging**